rpc-rs/src/rpc_client.rs

```diff
@@ -208,21 +256,86 @@ impl RpcClient {
 
         let topic = rpc_topic(&target, &self.lattice_prefix);
         let method = message.method.to_string();
-        let invocation = Invocation {
-            origin,
-            target,
-            operation: method.clone(),
-            msg: message.arg.into_owned(),
-            id: subject,
-            encoded_claims: claims.encode(&self.key).unwrap(),
-            host_id: self.host_id.clone(),
+        let len = message.arg.len();
+        let chunkify = {
+            cfg_if::cfg_if! {
+                if #[cfg(feature = "chunkify")] {
+                    crate::chunkify::needs_chunking(len)
+                } else {
+                    false
+                }
+            }
+        };
+
+        #[allow(unused_variables)]
+        let (invocation, body) = {
+            let mut inv = Invocation {
+                origin,
+                target,
+                operation: method.clone(),
+                id: subject,
+                encoded_claims: claims.encode(&self.key).unwrap(),
+                host_id: self.host_id.clone(),
+                content_length: Some(len as u64),
+                ..Default::default()
+            };
+            if chunkify {
+                (inv, Some(Vec::from(message.arg)))
+            } else {
+                inv.msg = Vec::from(message.arg);
+                (inv, None)
+            }
+        };
+        let nats_body = crate::common::serialize(&invocation)?;
+
+        let chunkify_join: Option<JoinHandle<RpcResult<()>>> = {
+            cfg_if::cfg_if! {
+                if #[cfg(feature = "chunkify")] {
+            if let Some(body) = body
+            {
+                let inv_id = invocation.id.clone();
+                debug!("chunkifying inv {} size {}", &inv_id, len);
+                // start chunking thread
+                let lattice = self.lattice_prefix().to_string();
+                Some(tokio::task::spawn_blocking(move || {
+                    let ce = chunkify_endpoint(None, lattice)?;
+                    ce.chunkify(&inv_id, &mut body.as_slice())
+                }))
+            } else {
+                None
+            }
+
+                    } else {
+                       None
+                    }
+                }
+        };
+        // I tried starting the send to ObjectStore in background thread,
+        // and then send the rpc, but if the objectstore hasn't completed,
+        // the recipient gets a missing object error,
+        // so we need to flush this first
```

Comment:
Yeah there's some opportunity to optimize for this in the future, but the host doesn't do this yet either - it completes the object store write before delivering the invocation.